compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs

```diff
@@ -331,6 +331,10 @@ isBadDependency x
     | Just (_ :: BadDependency) <- fromException x = True
     | otherwise = False
 
+-- | Catches BadDependency exception thrown by uses_ and use_
+-- and turns it into a Maybe.
+actionToMaybe :: Action a -> Action (Maybe a)
+actionToMaybe act = actionCatch (Just <$> act) (\(_ :: BadDependency) -> pure Nothing)
```

Comment:
It was useful to implement things like `getDefinition` where you want short circuiting. I’ve switched those over to use `MaybeT` + `use` now. Not entirely sure I like that given that we are now still mixing two different mechanisms for short circuiting but at least it’s somewhat consistent in the sense that we use `MaybeT` outside of rules and exceptions in a rule.

Side note: Turning that stuff into rules would mean that we can just use a single call to `use`. Leaving aside the fact that I think that change is orthogonal to this PR, it would also mean that the results would get cached which given the way we currently do garbage collection would mean that we gradually increase memory usage as users hover over more things so I don’t think turning it into rules is a good idea.

Reply:
I might be convinced its safe over time, and thus we can use it consistently, but at the moment I'm nervous about its semantics.

Agreed that turning this stuff into rules itself would be a bad idea.